transaction/std/src/memo/gift_code_cancellation.rs

```diff
@@ -0,0 +1,117 @@
+// Copyright (c) 2018-2022 The MobileCoin Foundation
+
+//! Object for 0x0202 Gift Code Cancellation memo type
+//!
+//! This was proposed for standardization in mobilecoinfoundation/mcips/pull/32
+
+use crate::{impl_memo_type_conversions, RegisteredMemoType};
+use std::convert::TryInto;
+
+/// Memo representing the cancellation of a gift code. If a gift code is
+/// never redeemed, the sender may cancel it by sending the TxOut back
+/// to their primary address. This memo will be written to the
+/// reserved change address with 8 bytes reserved for a u64 that
+/// represents the index of the cancelled gift code TxOut
+#[derive(Debug, Clone, Eq, PartialEq, Ord, PartialOrd)]
+pub struct GiftCodeCancellationMemo {
+    /// The data representing the gift code memo
+    memo_data: [u8; Self::MEMO_DATA_LEN],
+}
+
+impl RegisteredMemoType for GiftCodeCancellationMemo {
+    const MEMO_TYPE_BYTES: [u8; 2] = [0x02, 0x02];
+}
+
+impl GiftCodeCancellationMemo {
+    /// The length of the custom memo data
+    pub const MEMO_DATA_LEN: usize = 64;
+
+    /// Create a new gift code memo
+    pub fn new(global_index: u64) -> Self {
+        GiftCodeCancellationMemo::from(global_index)
+    }
+
+    /// Get the memo data
```

Comment:
So this is not the most useful API to expose here.

The `mc_transaction_core::MemoPayload` type gives the user access to the types. The `GiftCodeCancellationMemo` struct is supposed to be like a higher level type that abstracts away the details of the binary format.

So for instance, I would want to see like:

```
pub fn global_index(&self) -> u64;
pub fn note(&self) -> &str;
```

rather than just handing the user the entire Memo data, because that's not very useful to the SDKs and users, they would all have to refer to the MCIP to see how to interpret that data and we would get a ton of code duplication

Reply:
I see you have the `cancelled_gift_code_index` one, but I think we should get rid of the `memo_data` and make it just return the note, also checking that it's valid utf8

Reply:
The gift code cancellation txout actually doesn't have a utf-8 byte specification, in the MCIP - it's "unused". But I could remove the method which returns the bytes